File: src/client_beta1.cpp
#include <chrono>
#include <ctime>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <optional>
#include <regex>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

struct Profile {
    std::string video_codec = "libx265";
    int crf = 22;
    std::string preset = "medium";
    std::string audio_codec = "aac";
    std::string audio_bitrate = "192k";
    std::string container = "mkv";
};

struct Options {
    std::string input;
    std::string host;
    std::string user;
    int port = 22;
    std::string profile_path;
    std::string remote_base = "~/mkv_jobs";
    std::string output_dir = "./out";
    bool dry_run = false;
};

static std::string read_file(const std::string& path) {
    std::ifstream in(path);
    if (!in) throw std::runtime_error("Cannot open file: " + path);
    std::ostringstream ss;
    ss << in.rdbuf();
    return ss.str();
}

static std::optional<std::string> json_string(const std::string& text, const std::string& key) {
    std::regex r("\"" + key + "\"\\s*:\\s*\"([^\"]+)\"");
    std::smatch m;
    if (std::regex_search(text, m, r)) return m[1].str();
    return std::nullopt;
}

static std::optional<int> json_int(const std::string& text, const std::string& key) {
    std::regex r("\"" + key + "\"\\s*:\\s*([0-9]+)");
    std::smatch m;
    if (std::regex_search(text, m, r)) return std::stoi(m[1].str());
    return std::nullopt;
}

static Profile load_profile(const std::string& path) {
    const std::string raw = read_file(path);
    Profile p;
    if (auto v = json_string(raw, "video_codec")) p.video_codec = *v;
    if (auto v = json_int(raw, "crf")) p.crf = *v;
    if (auto v = json_string(raw, "preset")) p.preset = *v;
    if (auto v = json_string(raw, "audio_codec")) p.audio_codec = *v;
    if (auto v = json_string(raw, "audio_bitrate")) p.audio_bitrate = *v;
    if (auto v = json_string(raw, "container")) p.container = *v;
    return p;
}

static std::string shell_quote(const std::string& x) {
    std::string out = "'";
    for (char c : x) {
        if (c == '\'') out += "'\\''";
        else out += c;
    }
    out += "'";
    return out;
}

static int run_cmd(const std::string& cmd, bool dry_run) {
    std::cout << "$ " << cmd << "\n";
    if (dry_run) return 0;
    return std::system(cmd.c_str());
}

static std::string utc_job_id() {
    auto now = std::chrono::system_clock::now();
    std::time_t t = std::chrono::system_clock::to_time_t(now);
    std::tm* gmt = std::gmtime(&t);
    std::ostringstream os;
    os << "job_" << std::put_time(gmt, "%Y%m%dT%H%M%SZ");
    return os.str();
}

static void usage() {
    std::cout
        << "Usage: client_beta1_cpp <input.mkv> --host <host> --user <user> --profile <file.json> [options]\n"
        << "Options:\n"
        << "  --port <int>          SSH port (default: 22)\n"
        << "  --remote-base <dir>   Remote base dir (default: ~/mkv_jobs)\n"
        << "  --output-dir <dir>    Local output dir (default: ./out)\n"
        << "  --dry-run             Print commands only\n";
}

static Options parse_args(int argc, char** argv) {
    if (argc < 2) {
        usage();
        throw std::runtime_error("missing arguments");
    }

    Options o;
    o.input = argv[1];
    for (int i = 2; i < argc; ++i) {
        std::string a = argv[i];
        auto next = [&](const std::string& name) -> std::string {
            if (i + 1 >= argc) throw std::runtime_error("missing value for " + name);
            return argv[++i];
        };

        if (a == "--host") o.host = next(a);
        else if (a == "--user") o.user = next(a);
        else if (a == "--port") o.port = std::stoi(next(a));
        else if (a == "--profile") o.profile_path = next(a);
        else if (a == "--remote-base") o.remote_base = next(a);
        else if (a == "--output-dir") o.output_dir = next(a);
        else if (a == "--dry-run") o.dry_run = true;
        else throw std::runtime_error("unknown argument: " + a);
    }

    if (o.host.empty() || o.user.empty() || o.profile_path.empty()) {
        usage();
        throw std::runtime_error("required: --host --user --profile");
    }
    return o;
}

int main(int argc, char** argv) {
    try {
        Options opt = parse_args(argc, argv);

        std::filesystem::path in(opt.input);
        if (!std::filesystem::exists(in)) throw std::runtime_error("Input not found: " + opt.input);
        if (in.extension() != ".mkv") throw std::runtime_error("Input must be .mkv");
        if (!std::filesystem::exists(opt.profile_path)) throw std::runtime_error("Profile not found: " + opt.profile_path);

        Profile p = load_profile(opt.profile_path);
        std::filesystem::create_directories(opt.output_dir);

        const std::string job = utc_job_id();
        const std::string remote_dir = opt.remote_base + "/" + job;
        const std::string remote_input = remote_dir + "/" + in.filename().string();
        const std::string output_name = in.stem().string() + ".cppbeta1." + p.container;
        const std::string remote_output = remote_dir + "/" + output_name;
        const std::string local_output = (std::filesystem::path(opt.output_dir) / output_name).string();

        const std::string login = opt.user + "@" + opt.host;
        const std::string ssh_mkdir = "ssh -p " + std::to_string(opt.port) + " " + shell_quote(login) + " mkdir -p " + shell_quote(remote_dir);
        const std::string scp_upload = "scp -P " + std::to_string(opt.port) + " " + shell_quote(opt.input) + " " + shell_quote(login + ":" + remote_input);

        std::ostringstream ff;
        ff << "ffmpeg -y -threads 0 -i " << shell_quote(remote_input)
           << " -map 0 -c:v " << shell_quote(p.video_codec)
           << " -preset " << shell_quote(p.preset)
           << " -crf " << p.crf
           << " -c:a " << shell_quote(p.audio_codec)
           << " -b:a " << shell_quote(p.audio_bitrate)
           << " " << shell_quote(remote_output);

        const std::string ssh_ffmpeg = "ssh -p " + std::to_string(opt.port) + " " + shell_quote(login) + " " + ff.str();
        const std::string scp_download = "scp -P " + std::to_string(opt.port) + " " + shell_quote(login + ":" + remote_output) + " " + shell_quote(local_output);

        if (run_cmd(ssh_mkdir, opt.dry_run) != 0) throw std::runtime_error("mkdir command failed");
        if (run_cmd(scp_upload, opt.dry_run) != 0) throw std::runtime_error("upload failed");
        if (run_cmd(ssh_ffmpeg, opt.dry_run) != 0) throw std::runtime_error("ffmpeg failed");
        if (run_cmd(scp_download, opt.dry_run) != 0) throw std::runtime_error("download failed");

        std::cout << "Done: " << local_output << "\n";
        return 0;
    } catch (const std::exception& ex) {
        std::cerr << "Error: " << ex.what() << "\n";
        return 1;
    }
}

File: client_beta1.py
#!/usr/bin/env python3
"""Beta 1 CLI client for MKV Turbo Pipeline.

Flow:
1) Upload MKV from local machine to Ubuntu VDS via SCP
2) Run ffmpeg on VDS with a JSON profile
3) Download encoded file back to local machine
"""

from __future__ import annotations

import argparse
import json
import shlex
import subprocess
import sys
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path


@dataclass
class Profile:
    video_codec: str = "libx265"
    crf: int = 22
    preset: str = "medium"
    audio_codec: str = "aac"
    audio_bitrate: str = "192k"
    container: str = "mkv"

    @classmethod
    def load(cls, path: Path) -> "Profile":
        raw = json.loads(path.read_text(encoding="utf-8"))
        return cls(
            video_codec=raw.get("video_codec", cls.video_codec),
            crf=int(raw.get("crf", cls.crf)),
            preset=raw.get("preset", cls.preset),
            audio_codec=raw.get("audio_codec", cls.audio_codec),
            audio_bitrate=raw.get("audio_bitrate", cls.audio_bitrate),
            container=raw.get("container", cls.container),
        )


def run(command: list[str], dry_run: bool = False) -> None:
    printable = " ".join(shlex.quote(x) for x in command)
    print(f"$ {printable}")
    if dry_run:
        return

    result = subprocess.run(command, text=True)
    if result.returncode != 0:
        raise RuntimeError(f"Command failed ({result.returncode}): {printable}")


def make_ffmpeg_command(src: str, dst: str, profile: Profile) -> str:
    parts = [
        "ffmpeg",
        "-y",
        "-threads",
        "0",
        "-i",
        src,
        "-map",
        "0",
        "-c:v",
        profile.video_codec,
        "-preset",
        profile.preset,
        "-crf",
        str(profile.crf),
        "-c:a",
        profile.audio_codec,
        "-b:a",
        profile.audio_bitrate,
        dst,
    ]
    return " ".join(shlex.quote(x) for x in parts)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="MKV Turbo Beta 1 client")
    parser.add_argument("input", type=Path, help="Path to input .mkv")
    parser.add_argument("--host", required=True, help="VDS host")
    parser.add_argument("--user", required=True, help="SSH user")
    parser.add_argument("--port", type=int, default=22, help="SSH port")
    parser.add_argument("--profile", type=Path, required=True, help="Path to JSON profile")
    parser.add_argument("--remote-base", default="~/mkv_jobs", help="Remote base directory")
    parser.add_argument("--output-dir", type=Path, default=Path("./out"), help="Local output directory")
    parser.add_argument("--dry-run", action="store_true", help="Print commands without execution")
    return parser.parse_args()


def main() -> int:
    args = parse_args()

    if not args.input.exists():
        print(f"Input not found: {args.input}", file=sys.stderr)
        return 2
    if args.input.suffix.lower() != ".mkv":
        print("Input file must be .mkv", file=sys.stderr)
        return 2
    if not args.profile.exists():
        print(f"Profile not found: {args.profile}", file=sys.stderr)
        return 2

    profile = Profile.load(args.profile)
    args.output_dir.mkdir(parents=True, exist_ok=True)

    job_id = datetime.now(timezone.utc).strftime("job_%Y%m%dT%H%M%SZ")
    remote_dir = f"{args.remote_base.rstrip('/')}/{job_id}"
    remote_input = f"{remote_dir}/{args.input.name}"
    output_name = f"{args.input.stem}.beta1.{profile.container}"
    remote_output = f"{remote_dir}/{output_name}"
    local_output = args.output_dir / output_name

    ssh_base = ["ssh", "-p", str(args.port), f"{args.user}@{args.host}"]

    try:
        run(ssh_base + [f"mkdir -p {shlex.quote(remote_dir)}"], dry_run=args.dry_run)
        run(["scp", "-P", str(args.port), str(args.input), f"{args.user}@{args.host}:{remote_input}"], dry_run=args.dry_run)

        ffmpeg = make_ffmpeg_command(remote_input, remote_output, profile)
        run(ssh_base + [ffmpeg], dry_run=args.dry_run)

        run(["scp", "-P", str(args.port), f"{args.user}@{args.host}:{remote_output}", str(local_output)], dry_run=args.dry_run)
    except RuntimeError as err:
        print(str(err), file=sys.stderr)
        return 1

    print(f"Done: {local_output}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
